ICAFIX/scripts/xgboost_ensemble_classifier.py

```diff
@@ -0,0 +1,146 @@
+            else:
+                y_0_random_indices = np.random.choice(y_0_indices, size=int(ratio*len(y_0_indices)), replace=self.replace)
+                y_1_random_indices = np.random.choice(y_1_indices, size=int(ratio*len(y_0_indices)), replace=self.replace)
+                
+            # Combine the indices of y=0 and sampled y=1
+            balanced_indices = np.concatenate((y_0_random_indices, y_1_random_indices))
+
+            # Create the balanced feature matrix and label vector
+            balanced_X = X[balanced_indices]
+            balanced_y = y[balanced_indices]
+
+            model.fit(balanced_X, balanced_y)
+
+        self.is_fitted_ = True
+        # Return the classifier
+        return self
+    
+    def predict_proba(self, X):
+
+        # Check is fit had been called
+        assert self.is_fitted_ is True, 'model is not fitted'
+
+        # Input validation
+        X = check_array(X)
+        
+        y_hat = np.zeros((X.shape[0], 2))
+        for model in self.models:
+            y_hat[:,0] += model.predict_proba(X)[:,0]
+            y_hat[:,1] += model.predict_proba(X)[:,1]
+        y_hat = y_hat / len(self.models)
+        
+        return y_hat
+    
+    def predict(self, X):
+
+        # Check is fit had been called
+        assert self.is_fitted_ is True, 'model is not fitted'
+
+        # Input validation
+        X = check_array(X)
+        
+        predictions_proba=self.predict_proba(X)
+        
+        return predictions_proba[:,1]>=self.threshold
+    
+    def save(self, save_path):
+        # Check is fit had been called
+        assert self.is_fitted_ is True, 'model is not fitted'
+        
+        for i,model in enumerate(self.models):
+            model.save_model(save_path+f'/xgb_classifier_{i}.json')
+        print("model is saved!")
+        # Return the classifier
+        return self
+    
+    def load(self, model_path):
+        for i,model in enumerate(self.models):
+            model.load_model(model_path+f'/xgb_classifier_{i}.json')
+        self.is_fitted_ = True
+        print("model is loaded!")
+        # Return the classifier
+        return self
+
+# directly run is not supported because of the relative module path
+if __name__=='__main__':
+    from sklearn.datasets import load_breast_cancer
+    data = load_breast_cancer()
+    mdl=XGBoostEnsembleClassifier(random_state=12345)
+    mdl.fit(data.data, data.target)
+    pred_0=mdl.predict(data.data)
+    pred_0_1=mdl.predict_proba(data.data)
+    
+    mdl=XGBoostEnsembleClassifier(random_state=12345)
+    mdl.fit(data.data, data.target)
+    pred_1=mdl.predict(data.data)
+    pred_1_1=mdl.predict_proba(data.data)
+    
+    mdl=XGBoostEnsembleClassifier(random_state=0)
+    mdl.fit(data.data, data.target)
+    pred_2=mdl.predict(data.data)
+    pred_2_1=mdl.predict_proba(data.data)
+    
+    save_path = '/media/myelin/alex/tICA/tICAClassify/results/test'
```

Comment:
We should remove this hardcoded path on our filesystem one way or another.  This looks like it is saving a standard demo on a standard breast cancer dataset, and therefore all this code isn't useful, and can be deleted (rather than turning the path into an argument).